File: sidecar/src/routes/lambdaRoutes.ts
import { Router, type Request, type Response } from 'express';
import { AwsCliError } from '../infrastructure/awsCli';
import { LambdaService } from '../application/lambdaService';

type AsyncHandler = (request: Request, response: Response) => Promise<unknown>;

export const createLambdaRouter = (lambdaService: LambdaService) => {
  const router = Router();

  router.get('/lambda/capabilities', handle(async (_request, response) => {
    response.json(await lambdaService.getCapabilities());
  }));

  router.get('/lambda/functions', handle(async (_request, response) => {
    response.json(await lambdaService.listFunctions());
  }));

  router.post('/lambda/functions', handle(async (request, response) => {
    response.status(201).json(await lambdaService.createFunction(request.body));
  }));

  router.put('/lambda/functions/:functionName/code', handle(async (request, response) => {
    response.json(await lambdaService.updateFunctionCode(request.params.functionName, request.body));
  }));

  router.put('/lambda/functions/:functionName/configuration', handle(async (request, response) => {
    response.json(await lambdaService.updateFunctionConfiguration(request.params.functionName, request.body));
  }));

  router.post('/lambda/functions/:functionName/invoke', handle(async (request, response) => {
    response.json(await lambdaService.invokeFunction(request.params.functionName, request.body));
  }));

  router.get('/lambda/functions/:functionName/logs', handle(async (request, response) => {
    response.json(await lambdaService.getLogs(request.params.functionName));
  }));

  router.delete('/lambda/functions/:functionName', handle(async (request, response) => {
    response.json(await lambdaService.deleteFunction(request.params.functionName));
  }));

  return router;
};

const handle = (handler: AsyncHandler) => async (request: Request, response: Response) => {
  try {
    await handler(request, response);
  } catch (error) {
    const status = error instanceof AwsCliError ? 502 : 400;
    response.status(status).json({
      error: error instanceof Error ? error.message : String(error),
      stderr: error instanceof AwsCliError ? error.stderr : undefined,
    });
  }
};

File: mcp/floci_backend/api/lambda_routes.py
from fastapi import APIRouter, HTTPException
from typing import Dict, Any

from floci_backend.application.lambda_service import LambdaService

def create_lambda_router(lambda_service: LambdaService) -> APIRouter:
    router = APIRouter()

    @router.get('/lambda/capabilities')
    async def get_capabilities() -> Dict[str, Any]:
        return await lambda_service.get_capabilities()

    @router.get('/lambda/functions')
    async def list_functions() -> Dict[str, Any]:
        result = await lambda_service.list_functions()
        return {'ok': True, **result}

    @router.post('/lambda/functions')
    async def create_function(request: Dict[str, Any]) -> Dict[str, Any]:
        try:
            result = await lambda_service.create_function(request)
            return {'ok': True, **result}
        except Exception as error:
            raise HTTPException(status_code=500, detail=str(error))

    @router.put('/lambda/functions/{function_name}/code')
    async def update_function_code(function_name: str, request: Dict[str, Any]) -> Dict[str, Any]:
        try:
            result = await lambda_service.update_function_code(function_name, request)
            return {'ok': True, **result}
        except Exception as error:
            raise HTTPException(status_code=500, detail=str(error))

    @router.put('/lambda/functions/{function_name}/configuration')
    async def update_function_configuration(function_name: str, request: Dict[str, Any]) -> Dict[str, Any]:
        try:
            result = await lambda_service.update_function_configuration(function_name, request)
            return {'ok': True, **result}
        except Exception as error:
            raise HTTPException(status_code=500, detail=str(error))

    @router.post('/lambda/functions/{function_name}/invoke')
    async def invoke_function(function_name: str, request: Dict[str, Any]) -> Dict[str, Any]:
        try:
            result = await lambda_service.invoke_function(function_name, request)
            return {'ok': True, **result}
        except Exception as error:
            raise HTTPException(status_code=500, detail=str(error))

    @router.delete('/lambda/functions/{function_name}')
    async def delete_function(function_name: str) -> Dict[str, Any]:
        try:
            result = await lambda_service.delete_function(function_name)
            return {'ok': True, **result}
        except Exception as error:
            raise HTTPException(status_code=500, detail=str(error))

    @router.get('/lambda/functions/{function_name}/logs')
    async def get_logs(function_name: str) -> Dict[str, Any]:
        try:
            result = await lambda_service.get_logs(function_name)
            return {'ok': True, **result}
        except Exception as error:
            raise HTTPException(status_code=500, detail=str(error))

    return router
